### PR TITLE
Protected pages

### PR DESCRIPTION
**Pull request by AntonioGargaro**
All routes (pages) starting with `/admin` on client now require authorization based on `jwt_token`.

On every request, the Authorization header is automatically attached so you **DON'T** need to add it manually.

### CODE REVIEW

File: client/src/components/Main/Main.jsx [hunk 1]
```diff
@@ -12,10 +18,83 @@ import SurveyBuilderFromTemplate from '../../pages/SurveyBuilderFromTemplate';
...
+        render={(props) => {
+          return !checkingIfAuthed && auth
+          ? <Component {...props} />
+          : <LoadingPageOrRedirect {...props} checkingIfAuthed={checkingIfAuthed} />
+        }}
+      />
+      )
+  }
+
+  const routes = [
```
Comment (Joaoviana) [inline client/src/components/Main/Main.jsx]: would be great to save routes and protectedRoutes as an array of objects. And then map the values in each obj and render a ProtectedRoute / Route component. please check CreateSurvey.jsx on how I render the `fields` array

File: client/src/utils/addAuthorizationHeaderToEveryRequest.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import axios from 'axios';
+
+export default () => {
+  const token = window.localStorage.getItem('jwt_token');
+  if (token) {
+    axios.defaults.headers.common.Authorization = `JWT ${token}`;
```
Comment (TKostrzewski) [inline client/src/utils/addAuthorizationHeaderToEveryRequest.js]: oooh spicy

File: client/src/utils/deleteTokenOn401StatusCodes.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  };
+
+  const errorHandler = (error) => {
+    if (error.response.status === 401) {
+      window.localStorage.removeItem('jwt_token');
+    }
+    return Promise.reject(error);
+  };
+
+  axios.interceptors.response.use(
```
Comment (TKostrzewski) [inline client/src/utils/deleteTokenOn401StatusCodes.js]: extra spice

File: client/src/utils/addAuthorizationHeaderToEveryRequest.js (hunk 2 above)
Comment (AntonioGargaro) [reply client/src/utils/addAuthorizationHeaderToEveryRequest.js]: Thought you might like dat

File: client/src/utils/addAuthorizationHeaderToEveryRequest.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import axios from 'axios';
+
+export default () => {
+  const token = window.localStorage.getItem('jwt_token');
+  if (token) {
```
Comment (Joaoviana) [inline client/src/utils/addAuthorizationHeaderToEveryRequest.js]: use ternary op
 axios.defaults.headers.common.Authorization = token ? token : null

File: client/src/utils/deleteTokenOn401StatusCodes.js (hunk 3 above)
Comment (AntonioGargaro) [reply client/src/utils/deleteTokenOn401StatusCodes.js]: triple flame

File: client/src/utils/checkTokenIsAuth.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      const payload = {
+        auth: true,
+      };
+      dispatch({ type: SET_LOGIN, payload });
+      return true;
+    }
+  } catch (err) {
+    const res = err.response;
+
+    if (res && res.status === 401 && currentAuth === true) {
```
Comment (Joaoviana) [inline client/src/utils/checkTokenIsAuth.js]: can you save this as a more declarative variable

File: client/src/components/Main/Main.jsx (hunk 1 above)
Comment (Joaoviana) [reply client/src/components/Main/Main.jsx]: believe it is inside CreateSurvey/ and index.jsx

Comment (TKostrzewski) [reply client/src/components/Main/Main.jsx]: Interesting idea

File: client/src/components/Main/dummyData/dummyData.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
+export default [
```
Comment (thomas-galligan) [inline client/src/components/Main/dummyData/dummyData.js]: why do we need this file? Is the other dummy data not sufficient?

File: client/src/components/Main/Main.jsx [hunk 7]
```diff
@@ -1,21 +1,61 @@
...
+import { CHECKING_IF_AUTHED } from '../../store/actions/mainActions';
+
+const Main = ({ history }) => {
+  const dispatch = useDispatch();
 
-const Main = () => {
   const { data } = useSelector((state) => state.adminLoginReducer);
   const { auth } = data;
   const { snackbar } = useSelector((state) => state.snackbarReducer);
+  const { checkingIfAuthed } = useSelector((state) => state.mainReducer);
```
Comment (thomas-galligan) [inline client/src/components/Main/Main.jsx]: given this isn't a function I don't think its name should be a verb. Can we just call it `userAuthorised` or similar?

File: client/src/store/reducer/mainReducer.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { CHECKING_IF_AUTHED } from '../actions/mainActions';
+
+const initalState = {
+  checkingIfAuthed: true,
+};
+
+const adminLoginReducer = (state = initalState, action) => {
```
Comment (thomas-galligan) [inline client/src/store/reducer/mainReducer.js]: Given that this file exports adminLoginReducer, can the file have the same name?

File: client/src/store/actions/mainActions.js [hunk 9]
```diff
@@ -0,0 +1,2 @@
+/* eslint-disable import/prefer-default-export */
+export const CHECKING_IF_AUTHED = 'CHECKING_IF_AUTHED';
```
Comment (thomas-galligan) [inline client/src/store/actions/mainActions.js]: `mainActions` seems a bit grandiose ;)

File: client/src/components/Main/Main.jsx [hunk 10]
```diff
@@ -1,21 +1,61 @@
...
+        render={(props) => {
+          return !checkingIfAuthed && auth
+          ? <Component {...props} />
+          : <LoadingPageOrRedirect {...props} checkingIfAuthed={checkingIfAuthed} />
+        }}
+      />
+      )
+  }
+
+  const routesMap = routes.map(rProps => <Route key={rProps.path} {...rProps} />)
```
Comment (thomas-galligan) [inline client/src/components/Main/Main.jsx]: what does `rProps` mean? route props?

File: client/src/pages/AdminLogin/index.jsx [hunk 11]
```diff
@@ -16,31 +20,37 @@ import {
...
+
+  const { from } = location.state || { from: { pathname: "/admin" } };
 
   const { helperText, data } = useSelector((state) => state.adminLoginReducer);
   const { username, password } = data;
 
+  useEffect(() => {
+    const { auth } = data;
+    if (auth){
+      setTimeout(() => history.replace(from), 200);
```
Comment (thomas-galligan) [inline client/src/pages/AdminLogin/index.jsx]: Why are you using `setTimeout` here?

File: client/src/components/LoadingWheel/index.jsx [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import styles from './LoadingWheel.module.css';
+
+const LoadingWheel = () => (
+  <div id={styles['loader-container']}>
+    <div className={styles.loader} />
+  </div>
+);
```
Comment (thomas-galligan) [inline client/src/components/LoadingWheel/index.jsx]: bit late now but there are some lovely loading wheels part of Material UI: https://material-ui.com/components/progress/

File: client/src/utils/deleteTokenOn401StatusCodes.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+export default (axios) => {
+  const successHandler = (response) => {
+    return response;
+  };
```
Comment (thomas-galligan) [inline client/src/utils/deleteTokenOn401StatusCodes.js]: wat

Comment (thomas-galligan) [reply client/src/utils/deleteTokenOn401StatusCodes.js]: if you really must have this function, then you can do `const successHandler = (response) => response`

File: client/src/utils/deleteTokenOn401StatusCodes.js (hunk 3 above)
Comment (thomas-galligan) [reply client/src/utils/deleteTokenOn401StatusCodes.js]: I don't understand this but it does indeed look spicy

File: server/config/passport/index.js [hunk 14]
```diff
@@ -18,7 +18,6 @@ module.exports = (passport) => {
       },
       async (username, password, done) => {
         try {
-          // TODO not working currently
```
Comment (thomas-galligan) [inline server/config/passport/index.js]: lovely stuff

File: client/src/components/Main/Main.jsx (hunk 1 above)
Comment (AntonioGargaro) [reply client/src/components/Main/Main.jsx]: Added `client/src/components/Main/routes.js` to accomplish this.

File: client/src/utils/checkTokenIsAuth.js (hunk 5 above)
Comment (AntonioGargaro) [reply client/src/utils/checkTokenIsAuth.js]: Agreed. Fixed!

File: client/src/components/Main/dummyData/dummyData.js (hunk 6 above)
Comment (AntonioGargaro) [reply client/src/components/Main/dummyData/dummyData.js]: There are multiple versions of dummy data defined as consts in test files all over client.
There is no central location on client to pull this data from. Maybe we should raise as an issue to refactor tests to take dummyData from a central point?

File: client/src/store/reducer/mainReducer.js (hunk 8 above)
Comment (AntonioGargaro) [reply client/src/store/reducer/mainReducer.js]: Apologies, adminLoginReducer is wrong. This is a reducer for the component `main`.
Thanks for spotting it!

File: client/src/store/actions/mainActions.js (hunk 9 above)
Comment (AntonioGargaro) [reply client/src/store/actions/mainActions.js]: Hmm agreed, although it it is actions for the `main` component! Any other suggestions?

File: client/src/components/Main/Main.jsx (hunk 10 above)
Comment (AntonioGargaro) [reply client/src/components/Main/Main.jsx]: That'd be correct!

File: client/src/pages/AdminLogin/index.jsx (hunk 11 above)
Comment (AntonioGargaro) [reply client/src/pages/AdminLogin/index.jsx]: Good spot! I think I was testing for appearances for 'fake async' as currently stuff would load to quickly - if that is a thing. 
Removed!

File: client/src/components/LoadingWheel/index.jsx (hunk 12 above)
Comment (AntonioGargaro) [reply client/src/components/LoadingWheel/index.jsx]: Good find - can be added in future and made as issue #52

File: client/src/utils/deleteTokenOn401StatusCodes.js (hunk 13 above)
Comment (AntonioGargaro) [reply client/src/utils/deleteTokenOn401StatusCodes.js]: Fixed.

File: server/config/passport/index.js (hunk 14 above)
Comment (AntonioGargaro) [reply server/config/passport/index.js]: ;D if you would notice correctly, it is red and has been removed ;D

File: client/src/store/actions/mainActions.js (hunk 9 above)
Comment (TKostrzewski) [reply client/src/store/actions/mainActions.js]: Just a note, most of the other reducers don't have an actions page associated with them. I have a feeling what you've done here is better practice so we should discuss how we bring in the actions into the reducer to keep consistency throughout.

Comment (AntonioGargaro) [reply client/src/store/actions/mainActions.js]: Agreed, made issue #59 for this.

**Pull request merged**